### PR TITLE
Vec assume

### PR DESCRIPTION
**Pull request by xukai92**
On working of https://github.com/yebai/Turing.jl/issues/255

I'm trying to vectorize the transformation as well (#276). I already done that for univariate and simplex.

TODOs
- [x] vec mat
- [x] vec mat transformation(s)

### CODE REVIEW

**Comment by xukai92** [comment]
Note about using array of array v.s. mat result from `vec_assume_mv.jl`:


```julia
# Test for vectorize UnivariateDistribution
@model vdemo() = begin
  phi = Vector{Vector{Real}}(N)
  phi ~ [Dirichlet(beta)]
end

t_vec = @elapsed res_vec = sample(vdemo(), alg)

@model vdemo() = begin
  phi = Matrix(2,N)
  phi ~ [Dirichlet(beta)]
end

t_vec_mat = @elapsed res_vec_mat = sample(vdemo(), alg)

@model vdemo() = begin
  phi = Vector{Vector{Real}}(N)
  for i = 1:N
    phi[i] ~ Dirichlet(beta)
  end
end

t_loop = @elapsed res = sample(vdemo(), alg)

println("Time for")
println("  Loop : $t_loop")
println("  Vec  : $t_vec")
println("  Vec2 : $t_vec_mat")
```

gives

```julia
Time for
  Loop : 3.33750112
  Vec  : 15.286217032
  Vec2 : 2.782706592
```

Mat seems much faster (should be also related to my implementation of converting mat back to nested array)

**Comment by xukai92** [comment]
Note: `Distributions.jl` doesn't support vectorisation version of `logpdf` of matrixvariate variables.

**Comment by xukai92** [comment]
Also there is a typo(bug) in their fake verctorization implementation. For this  reason I will not include the test file in runtest.jl

**Comment by yebai** [comment]
@xukai92 Do you want me to review this PR? Seems tests are currently broken.

**Comment by xukai92** [comment]
Seems there is a bug of the resuming MCMC feature which may cause numerical error very easily. I guess it might be the case that when a chain is resumed, I call the `find_good_eps` again, which should be disabled. Let me fix this first.

**Comment by xukai92** [comment]
@yebai I fixed few bugs and improve the stability a bit. Could you have a review now plz?

**Comment by yebai** [comment]
>Time for
>  Loop : 9.823776041
>  Vec  : 21.934951717
>  Vec2 : 6.140146129
> with N=20.
> Mat seems much faster (should be also related to my implementation of converting mat back to nested array)

 impressive speedup! I've started reviewing this PR and will add my comments soon.

File: src/samplers/hmc.jl [hunk 1]
```diff
@@ -152,18 +157,32 @@ assume{T<:Hamiltonian}(spl::Sampler{T}, dist::Distribution, vn::VarName, vi::Var
...
   vns = map(i -> copybyindex(vn, "[$i]"), 1:n)
 
-  rs = vi[vns]
+  rs = vi[vns]  # NOTE: inside Turing the Julia conversion should be sticked to
 
   acclogp!(vi, sum(logpdf(dist, rs, istrans(vi, vns[1]))))
 
-  rs
+  if isa(dist, UnivariateDistribution) || isa(dist, MatrixDistribution)
+    var = rs
```
Comment (yebai) [inline src/samplers/hmc.jl]: perhaps we should add a dimensionality assertion here:

```julia
@assert size(var) == size(rs)
```

File: src/samplers/hmc.jl [hunk 2]
```diff
@@ -152,18 +157,32 @@ assume{T<:Hamiltonian}(spl::Sampler{T}, dist::Distribution, vn::VarName, vi::Var
...
+  elseif isa(dist, MultivariateDistribution)
+    if isa(var, Vector)
+      for i = 1:n
+        var[i] = rs[:,i]
+      end
+    elseif isa(var, Matrix)
+      var = rs
```
Comment (yebai) [inline src/samplers/hmc.jl]: dimensionality assertion?

File: src/samplers/sampler.jl [hunk 3]
```diff
@@ -42,25 +42,42 @@ assume(spl::Void, dist::Distribution, vn::VarName, vi::VarInfo) = begin
...
   else
     rs = rand(dist, n)
-    for i = 1:n
-      push!(vi, vns[i], rs[i], dist, 0)
+
+    if isa(dist, UnivariateDistribution) || isa(dist, MatrixDistribution)
+      for i = 1:n
+        push!(vi, vns[i], rs[i], dist, 0)
+      end
+      var = rs
```
Comment (yebai) [inline src/samplers/sampler.jl]: dimensionality assertion?

File: src/samplers/sampler.jl [hunk 4]
```diff
@@ -42,25 +42,42 @@ assume(spl::Void, dist::Distribution, vn::VarName, vi::VarInfo) = begin
...
+    elseif isa(dist, MultivariateDistribution)
+      for i = 1:n
+        push!(vi, vns[i], rs[:,i], dist, 0)
+      end
+      if isa(var, Vector)
+        for i = 1:n
+          var[i] = rs[:,i]
+        end
+      elseif isa(var, Matrix)
+        var = rs
```
Comment (yebai) [inline src/samplers/sampler.jl]: dimensionality assertion?

File: src/samplers/support/transform.jl [hunk 5]
```diff
@@ -167,6 +198,23 @@ Distributions.logpdf{T}(d::SimplexDistribution, x::Vector{T}, transform::Bool) =
...
+      Z[k,:] = X[k,:] ./ (one(T) - sum(X[1:k-1,:],1))'
+    end
+    for k = 1:K-1
+      lp += log(Z[k,:]) + log(one(T) - Z[k,:]) + log(one(T) - sum(X[1:k-1,:], 1))'
+    end
+  end
+  lp
+end
+
+# REVIEW: why do we put this piece of code here?
```
Comment (yebai) [inline src/samplers/support/transform.jl]: the distributions package does not check `d.p[x] > 0.0`, so I override their definition here. It's probably better to submit a PR to Distributions package to fix this.

File: src/samplers/hmc.jl (hunk 1 above)
Comment (xukai92) [inline src/samplers/hmc.jl]: Yeah the assertions makes sense to me!

File: src/samplers/support/transform.jl (hunk 5 above)
Comment (xukai92) [inline src/samplers/support/transform.jl]: Yeah I created an issue (https://github.com/yebai/Turing.jl/issues/285) for this.

**Comment by xukai92** [comment]
Vec LDA 215s v.s. loop LDA 296s (https://travis-ci.org/yebai/Turing.jl/jobs/242001041)

**Comment by yebai** [comment]
> Vec LDA 215s v.s. loop LDA 296s (https://travis-ci.org/yebai/Turing.jl/jobs/242001041)
Nice speedup! 

We should probably cleanup this PR, fix the tests and merge it into the master branch now.

**Comment by xukai92** [comment]
Yes. Some recent commits fail on Travis because I disable the rescale of std in adaptation (we divided std by the `min(std)` before). The tests still pass on my local but they fail on Travis because of running too long for NUTS.

The reason I disable the re-scaling is because I observed that it will lead to very small accept rate & numerical error in some circumstances for HMCDA, mentioned here https://github.com/yebai/Turing.jl/issues/284.

Do you think re-scaling by mean may be good for both HMCDA and NUTS and make sense? I tried that on my local which looks fine and I just pushed another commit to see if it works on Travis.

**Comment by xukai92** [comment]
@yebai Tests on Travis look fine now. But tests on Apprvor got the same problem before - freeze before running actual tests. I'm playing around the settings/scripts to see if I can resolve it.

**Comment by xukai92** [comment]
NOTE: I did a profiling on the vectorized version of LDA model and it shows >70% time is spent by `gradient2()`.

**Comment by xukai92** [comment]
@yebai Can you give me permissions on appveyor so that I can terminate or restart tests? I have to wait for each one to be terminated by the sever to try to fix the issue, which is a bit slow-responsed.

**Comment by yebai** [comment]
Done - I've added you to appveyor permissions list.

**Comment by xukai92** [comment]
@yebai Appveyor actually freezes when executing `using Turing`. Any idea why this happens?

**Comment by yebai** [comment]
> @yebai Appveyor actually freezes when executing using Turing. Any idea why this happens?

`Turing` imports other packages when loaded. It's possible some packages being imported have errors/conflicts, most likely due to pre-compilation. To see these warnings or errors, try temporarily disable `@supress_err` function in Turing.

**Pull request merged**